### PR TITLE
Annotate System.Reflection.TypeExtensions for nullable references

### PR DESCRIPTION
**Pull request by danmoseley**
Contributes to #40623

### CODE REVIEW

File: src/System.Reflection.TypeExtensions/ref/System.Reflection.TypeExtensions.cs [hunk 1]
```diff
@@ -15,12 +15,12 @@ public static partial class AssemblyExtensions
     }
     public static partial class EventInfoExtensions
     {
-        public static System.Reflection.MethodInfo GetAddMethod(this System.Reflection.EventInfo eventInfo) { throw null; }
-        public static System.Reflection.MethodInfo GetAddMethod(this System.Reflection.EventInfo eventInfo, bool nonPublic) { throw null; }
-        public static System.Reflection.MethodInfo GetRaiseMethod(this System.Reflection.EventInfo eventInfo) { throw null; }
-        public static System.Reflection.MethodInfo GetRaiseMethod(this System.Reflection.EventInfo eventInfo, bool nonPublic) { throw null; }
-        public static System.Reflection.MethodInfo GetRemoveMethod(this System.Reflection.EventInfo eventInfo) { throw null; }
-        public static System.Reflection.MethodInfo GetRemoveMethod(this System.Reflection.EventInfo eventInfo, bool nonPublic) { throw null; }
+        public static System.Reflection.MethodInfo?  GetAddMethod(this System.Reflection.EventInfo eventInfo) { throw null; }
```
Comment (safern) [inline src/System.Reflection.TypeExtensions/ref/System.Reflection.TypeExtensions.cs]: Nit, this has an extra space after the `?` -- did you do it manually or used `GenAPI`?

File: src/System.Reflection.TypeExtensions/src/System/Reflection/TypeExtensions.cs [hunk 2]
```diff
@@ -177,25 +177,25 @@ public static PropertyInfo[] GetProperties(Type type, BindingFlags bindingAttr)
...
 
-        public static PropertyInfo GetProperty(Type type, string name, Type returnType)
+        public static PropertyInfo? GetProperty(Type type, string name, Type returnType)
         {
             Requires.NotNull(type, nameof(type));
             return type.GetProperty(name, returnType);
         }
 
-        public static PropertyInfo GetProperty(Type type, string name, Type returnType, Type[] types)
+        public static PropertyInfo? GetProperty(Type type, string name, Type returnType, Type[] types)
```
Comment (stephentoub) [inline src/System.Reflection.TypeExtensions/src/System/Reflection/TypeExtensions.cs]: GitHub won't let me comment on it, but Type.IsAssignableFrom accepts a `Type?`... we don't want to do the same with the below IsAssignableFrom?

Comment (stephentoub) [reply src/System.Reflection.TypeExtensions/src/System/Reflection/TypeExtensions.cs]: Same for IsInstanceOfType

File: src/System.Reflection.TypeExtensions/src/System/Reflection/TypeExtensions.cs [hunk 3]
```diff
@@ -177,25 +177,25 @@ public static PropertyInfo[] GetProperties(Type type, BindingFlags bindingAttr)
...
 
-        public static PropertyInfo GetProperty(Type type, string name, BindingFlags bindingAttr)
+        public static PropertyInfo? GetProperty(Type type, string name, BindingFlags bindingAttr)
         {
             Requires.NotNull(type, nameof(type));
             return type.GetProperty(name, bindingAttr);
         }
 
-        public static PropertyInfo GetProperty(Type type, string name, Type returnType)
+        public static PropertyInfo? GetProperty(Type type, string name, Type returnType)
```
Comment (safern) [inline src/System.Reflection.TypeExtensions/src/System/Reflection/TypeExtensions.cs]: It seems like `returnType` should be nullable. https://source.dot.net/#System.Private.CoreLib/shared/System/Type.cs,224

File: src/System.Reflection.TypeExtensions/src/System/Reflection/TypeExtensions.cs (hunk 2 above)
Comment (safern) [inline src/System.Reflection.TypeExtensions/src/System/Reflection/TypeExtensions.cs]: Ditto.

File: src/System.Reflection.TypeExtensions/src/System/Reflection/TypeExtensions.cs [hunk 4]
```diff
@@ -177,25 +177,25 @@ public static PropertyInfo[] GetProperties(Type type, BindingFlags bindingAttr)
...
         {
             Requires.NotNull(type, nameof(type));
             return type.GetProperty(name, returnType);
         }
 
-        public static PropertyInfo GetProperty(Type type, string name, Type returnType, Type[] types)
+        public static PropertyInfo? GetProperty(Type type, string name, Type returnType, Type[] types)
         {
             Requires.NotNull(type, nameof(type));
             return type.GetProperty(name, returnType, types);
```
Comment (safern) [inline src/System.Reflection.TypeExtensions/src/System/Reflection/TypeExtensions.cs]: GitHub won't let me comment but `IsAssignableFrom` accepts a null type for parameter c and returns false.

Comment (safern) [reply src/System.Reflection.TypeExtensions/src/System/Reflection/TypeExtensions.cs]: Also, `IsInstanceOfType` accepts a null `object`.

**Review by safern — APPROVED** [review]
Other than my comments, LGTM.

File: src/System.Reflection.TypeExtensions/src/System/Reflection/TypeExtensions.cs (hunk 2 above)
Comment (safern) [reply src/System.Reflection.TypeExtensions/src/System/Reflection/TypeExtensions.cs]: Ha, we commented at the same time 😄

File: src/System.Reflection.TypeExtensions/ref/System.Reflection.TypeExtensions.cs (hunk 1 above)
Comment (danmosemsft) [reply src/System.Reflection.TypeExtensions/ref/System.Reflection.TypeExtensions.cs]: I forgot about GenAPI. Remind me how to do it? I ran `msbuild /t:GenerateReferenceAssemblySource` in both ref and src and it didn't seem to do it.

Comment (danmosemsft) [reply src/System.Reflection.TypeExtensions/ref/System.Reflection.TypeExtensions.cs]: @buyaa-n ?

Comment (buyaa-n) [reply src/System.Reflection.TypeExtensions/ref/System.Reflection.TypeExtensions.cs]: it should work from src folder, also the option is little different: 
`msbuild /t:GenerateReferenceSource`

Comment (safern) [reply src/System.Reflection.TypeExtensions/ref/System.Reflection.TypeExtensions.cs]: Yeah that should do it.

**Comment by danmosemsft** [comment]
I do not understand why it is showing all these commits, when my branch is rebased and looks normal:
https://github.com/dotnet/corefx/compare/master...danmosemsft:S.R.TE
When you go there it shows this PR!
I'll open a new one.

**Pull request closed without merging**